postreise/plot/plot_sim_vs_hist.py

```diff
@@ -10,41 +11,66 @@ class PlotSettings:
     size_inches = (20, 10)
 
 
-def plot_generation_sim_vs_hist(sim_gen, hist_gen, s_info, state, showmax=True):
-    """
-    Plot simulated vs historical generation of each resource in the scenario
+def plot_generation_sim_vs_hist(
+    scenario,
+    hist_gen,
+    state,
+    show_max=True,
+    plot_show=True,
+):
+    """Plot simulated vs historical generation of each resource in the scenario
     for the given state. Optionally include max generation.
 
-    :param pandas.DataFrame sim_gen: simulated generation dataframe
-    :param pandas.DataFrame hist_gen: historical generation dataframe
-    :param powersimdata.design.ScenarioInfo s_info: scenario info instance.
+    :param powersimdata.scenario.scenario.Scenario scenario: scenario instance.
+    :param pandas.DataFrame hist_gen: historical generation data frame. Columns are
+        resources and indices are state names.
     :param str state: the US state
-    :param bool showmax: determine whether to additionally plot max generation of each resource
+    :param bool show_max: determine whether to additionally plot max generation of each
+        resource.
+    :param plot_show: show the plot or not, defaults to True.
+    :return: (*matplotlib.axes.Axes*) -- axes object of the plot.
+    :raises TypeError:
+        if ``hist_gen`` is not a data frame.
+        if ``state`` is not a str.
+    :raises ValueError: if ``state is not in ``hist_gen`` or ``scenario``.
     """
-    labels = list(sim_gen.columns)
-    all_resources = s_info.get_available_resource("all")
+    if not isinstance(hist_gen, pd.DataFrame):
+        raise TypeError("hist_gen must be a pandas.DataFrame")
+    if not isinstance(state, str):
+        raise TypeError("state must be a str")
+
+    sim_gen = sum_generation_by_state(scenario)
+    if state not in sim_gen.index or state not in hist_gen.index:
+        raise ValueError("Invalid state")
+
+    all_resources = list(sim_gen.columns)
 
     sim = [int(round(sim_gen.loc[state, res])) for res in all_resources]
     hist = [int(round(hist_gen.loc[state, res])) for res in all_resources]
 
-    def calc_max(gen_type):
-        with warnings.catch_warnings():
-            warnings.simplefilter("ignore")
-            return int(
-                s_info.get_capacity(gen_type, state) * len(s_info.pg.index) / 1000
-            )
+    def calc_max(fuel):
+        grid = scenario.get_grid()
+        loadzone = list(grid.model_immutables.area_to_loadzone(state, "state"))  # noqa
+        capacity = grid.plant.query("type == @fuel & zone_name == @loadzone")[
+            "Pmax"
+        ].sum()
+        if capacity == 0:
+            print(f"No {fuel} generator in {state}")
+
+        pg = scenario.get_pg()
+        return int(capacity * pg.shape[0] / 1000)
 
     max_gen = []
-    if showmax:
+    if show_max:
         max_gen = [calc_max(res) for res in all_resources]
```

Comment:
We could potentially refactor this to avoided repeated calls to `get_grid` and `get_pg` within `calc_max`, using a groupby on `type` for generators in the given zone, but I bet this way isn't too slow since we know that we don't have too many resources.

Reply:
Given `calc_max` is defined as an internal function, another option is imply making `grid` and `pg` available in the scope of the outer function, i.e. call `scenario.get_grid()` and `scenario.get_pg()` outside `calc_max`.

Reply:
I factor out `pg` and `grid` from the internal function as @BainanXia proposed